### PR TITLE
Kafka (sarama) authentication with user/pass

### PR DESCRIPTION
**Pull request by annanay25**
Adds authentication to connect to Kafka through the Sarama library.

Drop #6172 in favour of this PR.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by balamurugana — DISMISSED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by kannappanr** [comment]
@annanay25 Can you please resolve the conflicts 
@harshavardhana Can you PTAL after that

File: cmd/config-migrate.go [hunk 1]
```diff
@@ -2373,3 +2378,123 @@ func migrateV26ToV27() error {
 	logger.Info(configMigrateMSGTemplate, configFile, "26", "27")
 	return nil
 }
+
+func migrateV27ToV28() error {
```
Comment (harshavardhana) [inline cmd/config-migrate.go]: We don't have any migration code anymore we can avoid writing this.

File: pkg/event/target/kafka.go [hunk 2]
```diff
@@ -32,6 +33,16 @@ type KafkaArgs struct {
 	Enable  bool        `json:"enable"`
 	Brokers []xnet.Host `json:"brokers"`
 	Topic   string      `json:"topic"`
+	TLS     struct {
```
Comment (harshavardhana) [inline pkg/event/target/kafka.go:36]: just updating these fields is enough

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Config migration should be removed from this PR.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-tls.sh | :x: [more...](#b4c17361-62e6-4ed1-b3d3-57dbf31bc07f) |
| mint-gateway-azure.sh | :x: [more...](#6af1f27e-1b98-40d9-bc21-5cacd53a1e7b) |
| mint-large-bucket.sh | :x: [more...](#e4253318-17ed-47bf-afe3-3b75769ad788) |
<a name="e4253318-17ed-47bf-afe3-3b75769ad788"></a>
### 6291-79eccb3/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32751
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c1a9c75c9d15:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 57 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 56 seconds
(5/13) Running aws-sdk-ruby tests ... done in 18 seconds
(6/13) Running mc tests ... done in 32 seconds
(7/13) Running minio-dotnet tests ... done in 4 minutes and 32 seconds
(8/13) Running minio-go tests ... done in 1 minutes and 46 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 2 seconds
(10/13) Running minio-js tests ... done in 42 seconds
(11/13) Running minio-py tests ... FAILED in 1 hours 5 minutes and 3 seconds
{
  "duration": 3612507,
  "status": "FAIL",
  "name": "minio-py:test_thread_safe",
  "function": "put_object(bucket_name, object_name, data, length, content_type, metadata)",
  "args": {
    "object_name": "7025fb6b-519f-4b70-9902-dbc9be7198aa",
    "bucket_name": "minio-py-test-bfdbf669-c879-4117-b868-67d3c3faff75"
  },
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 600, in urlopen\n    chunked=chunked)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 384, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 2, in raise_from\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 380, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.5/http/client.py\", line 1197, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.5/http/client.py\", line 297, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.5/http/client.py\", line 266, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1427, in test_thread_safe\n    client.remove_object(bucket_name, object_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 942, in remove_object\n    object_name=object_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1752, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/poolmanager.py\", line 322, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 638, in urlopen\n    _stacktrace=sys.exc_info()[2])\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/util/retry.py\", line 398, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minikube', port=32751): Max retries exceeded with url: /minio-py-test-bfdbf669-c879-4117-b868-67d3c3faff75/7025fb6b-519f-4b70-9902-dbc9be7198aa (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1800, in main\n    test_thread_safe(client, testfile, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1430, in test_thread_safe\n    raise Exception(err)\nException: HTTPConnectionPool(host='minikube', port=32751): Max retries exceeded with url: /minio-py-test-bfdbf669-c879-4117-b868-67d3c3faff75/7025fb6b-519f-4b70-9902-dbc9be7198aa (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n",
  "message": "HTTPConnectionPool(host='minikube', port=32751): Max retries exceeded with url: /minio-py-test-bfdbf669-c879-4117-b868-67d3c3faff75/7025fb6b-519f-4b70-9902-dbc9be7198aa (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))"
}

Executed 10 out of 13 tests successfully.
```
<a name="6af1f27e-1b98-40d9-bc21-5cacd53a1e7b"></a>
### 6291-79eccb3/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31161
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4f36c93807af:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 40 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 6 seconds
(4/13) Running aws-sdk-php tests ... done in 8 minutes and 44 seconds
(5/13) Running aws-sdk-ruby tests ... done in 52 seconds
(6/13) Running mc tests ... done in 5 minutes and 36 seconds
(7/13) Running minio-dotnet tests ... done in 6 minutes and 3 seconds
(8/13) Running minio-go tests ... FAILED in 8 minutes and 39 seconds
{
  "args": {
    "policy": "{\"expiration\":\"2018-09-11T18:38:25.392Z\",\"conditions\":[[\"eq\",\"$bucket\",\"minio-go-test-o3lrbhvuer5amzuu\"],[\"eq\",\"$key\",\"d5lmadt50uh03f90rgmkzdydfdywmb\"],[\"eq\",\"$Content-Type\",\"binary/octet-stream\"],[\"eq\",\"$x-amz-meta-61bc06btwwhhowqgwy1atjvocu5g4z\",\"hphy3a340oemkwerzaf5o1qi4aq1bn\"],[\"content-length-range\", 10, 1048576]]}"
  },
  "duration": 381,
  "error": "400 Bad Request",
  "function": "PresignedPostPolicy(policy)",
  "message": "Http request failed",
  "name": "minio-go: testPresignedPostPolicy",
  "status": "FAIL"
}

Executed 7 out of 13 tests successfully.
```
<a name="b4c17361-62e6-4ed1-b3d3-57dbf31bc07f"></a>
### 6291-79eccb3/mint-tls.sh.log:
```
Running with
SERVER_ENDPOINT:      localhost
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         1
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2e34e73e5b32:/mint/log /tmp/mint-logs'
(1/13) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-4v5ruaci6dbmqt",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 397,
  "error": "RequestError: send request failed\ncaused by: Put https://localhost/aws-sdk-go-test-4v5ruaci6dbmqt: dial tcp 127.0.0.1:443: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```

**Comment by nitisht** [comment]
@annanay25 is this ready for review again?

**Comment by annanay25** [comment]
Hi @nitisht, yes, I've made the requested changes!

**Comment by nitisht** [comment]
> yes, I've made the requested changes!

ping @balamurugana @harshavardhana

**Review by balamurugana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**